// cobalt-uoft/lib/api/food/model.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _mongoose = require('mongoose');

var _mongoose2 = _interopRequireDefault(_mongoose);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Schema = _mongoose2.default.Schema;

var foodSchema = new Schema({
  id: String,
  building_id: String,
  name: String,
  short_name: String,
  description: String,
  url: String,
  tags: { type: [String], index: true },
  image: String,
  campus: String,
  lat: Number,
  lng: Number,
  address: String,
  hours: {
    saturday: {
      closed: Boolean,
      open: Number,
      close: Number
    },
    friday: {
      closed: Boolean,
      open: Number,
      close: Number
    },
    thursday: {
      closed: Boolean,
      open: Number,
      close: Number
    },
    wednesday: {
      closed: Boolean,
      open: Number,
      close: Number
    },
    tuesday: {
      closed: Boolean,
      open: Number,
      close: Number
    },
    monday: {
      closed: Boolean,
      open: Number,
      close: Number
    },
    sunday: {
      closed: Boolean,
      open: Number,
      close: Number
    }
  }
}, { collection: 'food' });

foodSchema.index({
  name: 'text',
  description: 'text',
  campus: 'text',
  tags: 'text'
});

exports.default = _mongoose2.default.model('food', foodSchema);
