// cobalt-uoft/lib/api/courses/routes/filter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

exports.default = filter;

var _model = require('../model');

var _model2 = _interopRequireDefault(_model);

var _co = require('co');

var _co2 = _interopRequireDefault(_co);

var _queryParser = require('../../utils/query-parser');

var _queryParser2 = _interopRequireDefault(_queryParser);

var _filterMapReduce = require('./filterMapReduce');

var _filterMapReduce2 = _interopRequireDefault(_filterMapReduce);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Mapping of valid filter keys
var KEYMAP = {
  'code': { type: 'string', value: 'code' },
  'name': { type: 'string', value: 'name' },
  'description': { type: 'string', value: 'description' },
  'department': { type: 'string', value: 'department' },
  'division': { type: 'string', value: 'division' },
  'prerequisite': { type: 'string', value: 'prerequisites' },
  'exclusion': { type: 'string', value: 'exclusions' },
  'level': { type: 'number', value: 'level' },
  'breadth': { type: 'number', value: 'breadths' },
  'campus': { type: 'string', value: 'campus' },
  'term': { type: 'string', value: 'term' },
  'meeting_code': { type: 'string', value: 'meeting_sections.code', relativeValue: 'code' },
  'instructor': { type: 'string', value: 'meeting_sections.instructors', relativeValue: 'instructors' },
  'day': { type: 'string', value: 'meeting_sections.times.day', relativeValue: 'day' },
  'start': { type: 'time', value: 'meeting_sections.times.start', relativeValue: 'start' },
  'end': { type: 'time', value: 'meeting_sections.times.end', relativeValue: 'end' },
  'duration': { type: 'time', value: 'meeting_sections.times.duration', relativeValue: 'duration' },
  'location': { type: 'string', value: 'meeting_sections.times.location', relativeValue: 'location' },
  'size': { type: 'number', value: 'meeting_sections.size', relativeValue: 'size' },
  'enrolment': { type: 'number', value: 'meeting_sections.enrolment', relativeValue: 'enrolment' }
};

function filter(req, res, next) {
  // Generate parsed tokens and filters from query
  var query = _queryParser2.default.parseQuery(req.query.q, KEYMAP);

  if (query.mapReduce) {
    (0, _co2.default)(_regenerator2.default.mark(function _callee() {
      var docs, i;
      return _regenerator2.default.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.prev = 0;
              _context.next = 3;
              return _model2.default.mapReduce({
                query: query.filter,
                scope: {
                  q: query.tokens,
                  keyMap: KEYMAP
                },
                limit: req.query.limit,
                map: _filterMapReduce2.default.map,
                reduce: _filterMapReduce2.default.reduce
              });

            case 3:
              docs = _context.sent;

              for (i = 0; i < docs.length; i++) {
                docs[i] = docs[i].value;
              }
              res.json(docs);
              _context.next = 11;
              break;

            case 8:
              _context.prev = 8;
              _context.t0 = _context['catch'](0);
              return _context.abrupt('return', next(_context.t0));

            case 11:
            case 'end':
              return _context.stop();
          }
        }
      }, _callee, this, [[0, 8]]);
    }));
  } else {
    (0, _co2.default)(_regenerator2.default.mark(function _callee2() {
      var docs;
      return _regenerator2.default.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _context2.prev = 0;
              _context2.next = 3;
              return _model2.default.find(query.filter, '-__v -_id -meeting_sections._id -meeting_sections.times._id').limit(req.query.limit).skip(req.query.skip).sort(req.query.sort).exec();

            case 3:
              docs = _context2.sent;

              res.json(docs);
              _context2.next = 10;
              break;

            case 7:
              _context2.prev = 7;
              _context2.t0 = _context2['catch'](0);
              return _context2.abrupt('return', next(_context2.t0));

            case 10:
            case 'end':
              return _context2.stop();
          }
        }
      }, _callee2, this, [[0, 7]]);
    }));
  }
}
